drivers/sensor/b57861s/b57861s.c

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright (c) 2023 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT epcos_b57861s
+
+#include <zephyr/drivers/adc.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(B57861S, CONFIG_SENSOR_LOG_LEVEL);
+
+struct ntc_compensation {
+	int8_t temp_c;
+	uint32_t ohm;
+};
+
+/*
+ * A compensation table should be sorted by the values of .ohm
+ * in descending order.
+ * The following compensation tables are from the specifications in EPCOS NTC
+ * Thermistors Datasheets
+ * See - https://www.tdk-electronics.tdk.com/web/designtool/ntc/
+ */
+
+/* clang-format off */
+static const struct ntc_compensation b57861s[] = {
+	{ .temp_c =  -40, .ohm = 336500	},
+	{ .temp_c =  -35, .ohm = 242590	},
+	{ .temp_c =  -30, .ohm = 177000	},
+	{ .temp_c =  -25, .ohm = 130370	},
+	{ .temp_c =  -20, .ohm =  97070	},
+	{ .temp_c =  -15, .ohm =  72929	},
+	{ .temp_c =  -10, .ohm =  55330	},
+	{ .temp_c =   -5, .ohm =  42315	},
+	{ .temp_c =    0, .ohm =  32650	},
+	{ .temp_c =    5, .ohm =  25388	},
+	{ .temp_c =   10, .ohm =  19900	},
+	{ .temp_c =   15, .ohm =  15708	},
+	{ .temp_c =   20, .ohm =  12490	},
+	{ .temp_c =   25, .ohm =  10000	},
+	{ .temp_c =   30, .ohm =   8057	},
+	{ .temp_c =   35, .ohm =   6531	},
+	{ .temp_c =   40, .ohm =   5327	},
+	{ .temp_c =   45, .ohm =   4368	},
+	{ .temp_c =   50, .ohm =   3603	},
+	{ .temp_c =   55, .ohm =   2986	},
+	{ .temp_c =   60, .ohm =   2488	},
+	{ .temp_c =   65, .ohm =   2083	},
+	{ .temp_c =   70, .ohm =   1752	},
+	{ .temp_c =   75, .ohm =   1481	},
+	{ .temp_c =   80, .ohm =   1258	},
+	{ .temp_c =   85, .ohm =   1072	},
+	{ .temp_c =   90, .ohm =    917	},
+	{ .temp_c =   95, .ohm =    788	},
+	{ .temp_c =  100, .ohm =    680	},
+	{ .temp_c =  105, .ohm =    588	},
+	{ .temp_c =  110, .ohm =    511	},
+	{ .temp_c =  115, .ohm =    445	},
+	{ .temp_c =  120, .ohm =    389	},
+	{ .temp_c =  125, .ohm =    341	},
+};
+/* clang-format on */
+
+struct ntc_type {
+	const struct ntc_compensation *comp;
+	int n_comp;
+};
+
+static const struct ntc_type b57861s_type = {
+	.comp = b57861s,
+	.n_comp = ARRAY_SIZE(b57861s),
+};
+
+struct b57861s_sensor_data {
+	const struct ntc_type *comp_table;
+	struct k_mutex mutex;
+	int16_t raw;
+	int16_t sample_val;
+};
+
+enum ntc_type_e {
+	NTC_CONNECTED_POSITIVE,
+	NTC_CONNECTED_GROUND
+};
+
+struct b57861s_sensor_config {
+	const struct adc_dt_spec adc_channel;
+	uint32_t r25_ohm;
+	uint32_t pullup_mv;
+	uint32_t pullup_ohm;
+	uint32_t pulldown_ohm;
+	enum ntc_type_e connection_type;
+};
+
+static void lookup_comp(const struct ntc_type *data, unsigned int ohm, int *i_low, int *i_high)
```

Comment:
Could `bsearch()` help here?

Reply:
`bsearch` looks to find an exact match and returns a pointer to it, if not found returns NULL.  
While `lookup_comp` performs a binary search, its most unlikely to find an ohm value that matches.  It finds the `low` and `high` indexes where `ohm` falls between.  I might be able to make the compare function passed to `bsearch` to return found in these scenarios but seems a little hacky to me as will need to check the pointer returned then check if it's upper or lower bound.

Reply:
https://pastebin.com/prpRVmjn

Reply:
That's cool. If you would consider another generic tweak, you could put the table into Device Tree 😃